schemachanger: implement ALTER COLUMN [SET | DROP] ON UPDATE in declarative

rafiss opened the pull request:
This adds support for setting and dropping ON UPDATE expressions to the
declarative schema changer. No new elements were needed, so this was
pretty straightforward.


fixes https://github.com/cockroachdb/cockroach/issues/139604

Release note: None

rafiss committed:
schemachanger: implement ALTER COLUMN [SET | DROP] ON UPDATE in declarative

This adds support for setting and dropping ON UPDATE expressions to the
declarative schema changer. No new elements were needed, so this was
pretty straightforward.

Release note: None

rafiss committed:
schemachanger: add builder and planner tests for DEFAULT / ON UPDATE

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149444)
<!-- Reviewable:end -->

rafiss committed:
sql: standardize error message for ON UPDATE expressions

This makes the legacy schema changer return the same error text that the
declarative one has.

Release note: None

spilchen reviewed (commented):
> Reviewed 21 of 23 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149444)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_on_update.go` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149444#-OUushmz9hwHdPHdC6CM:-OUushmz9hwHdPHdC6CN:b-pblphl) ([raw file](https://github.com/cockroachdb/cockroach/blob/63c5ba3c68611677e053d5f5cfe35986725679f0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_on_update.go#L13)):*
> ```Go
> )
> 
> func alterTableSetOnUpdate(
> ```

Where is the check to verify there are no FKs that have an ON UPDATE clause? We have this for the legacy schema changer:
https://github.com/rafiss/cockroach/blob/9ca3788a1fdcc29a3ab2442f5c034f585267faee/pkg/sql/alter_table.go#L996-L1006

I couldn't see where this is checked for DSC. But I see it's tested in the `on_update` logic test, which passes for you. So, maybe I'm missing how this check is applied.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149444)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_on_update.go` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149444#-OUushmz9hwHdPHdC6CM:-OV8NVX28gsLJ8xoVds5:b-rwl4fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/63c5ba3c68611677e053d5f5cfe35986725679f0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_on_update.go#L13)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Where is the check to verify there are no FKs that have an ON UPDATE clause? We have this for the legacy schema changer:
https://github.com/rafiss/cockroach/blob/9ca3788a1fdcc29a3ab2442f5c034f585267faee/pkg/sql/alter_table.go#L996-L1006

I couldn't see where this is checked for DSC. But I see it's tested in the `on_update` logic test, which passes for you. So, maybe I'm missing how this check is applied.
</blockquote></details>

the DSC does this check during the descriptor validation that we execute when calling `(*descs.Collection).WriteDescToBatch`:

https://github.com/cockroachdb/cockroach/blob/606cc0df0209a57a25343740bd9982bb35df40cc/pkg/sql/catalog/tabledesc/validate.go#L1144-L1147


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 23 files at r1, 5 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149444#-:-OV8wuIg5xw-wbjVChIt:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_on_update.go` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149444#-OUushmz9hwHdPHdC6CM:-OV8wKG88YoESRol5nfs:b-7lz0p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c79a5403f56e958b15eac0e716bb4ab09c6623/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_on_update.go#L13)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the DSC does this check during the descriptor validation that we execute when calling `(*descs.Collection).WriteDescToBatch`:

https://github.com/cockroachdb/cockroach/blob/606cc0df0209a57a25343740bd9982bb35df40cc/pkg/sql/catalog/tabledesc/validate.go#L1144-L1147
</blockquote></details>

Ah ok. I figured I missed something. Thanks for clarifying.


<!-- Sent from Reviewable.io -->
